source/reduce/structured_loop_to_selection_reduction_opportunity.cpp

```diff
@@ -0,0 +1,406 @@
+      {{loop_merge_inst->GetOperand(kMergeNodeIndex).type,
+        {loop_merge_block_id}},
+       {SPV_OPERAND_TYPE_SELECTION_CONTROL, {SpvSelectionControlMaskNone}}});
+
+  // The loop header either finishes with OpBranch or OpBranchConditional.
+  // The latter is fine for a selection.  In the former case we need to turn
+  // it into OpBranchConditional.  We use "true" as the condition, and make
+  // the "else" branch be the merge block.
+  auto terminator = loop_construct_header_->terminator();
+  if (terminator->opcode() == SpvOpBranch) {
+    analysis::Bool temp;
+    const analysis::Bool* bool_type =
+        context_->get_type_mgr()->GetRegisteredType(&temp)->AsBool();
+    auto const_mgr = context_->get_constant_mgr();
+    auto true_const = const_mgr->GetConstant(bool_type, {true});
+    auto true_const_result_id =
+        const_mgr->GetDefiningInstruction(true_const)->result_id();
+    auto original_branch_id = terminator->GetSingleWordOperand(0);
+    terminator->SetOpcode(SpvOpBranchConditional);
+    terminator->ReplaceOperands({{SPV_OPERAND_TYPE_ID, {true_const_result_id}},
+                                 {SPV_OPERAND_TYPE_ID, {original_branch_id}},
+                                 {SPV_OPERAND_TYPE_ID, {loop_merge_block_id}}});
+    if (original_branch_id != loop_merge_block_id) {
+      // TODO(afd): consider adding a test for the case where they are equal.
+      AdaptPhiNodesForAddedEdge(loop_construct_header_->id(),
+                                context_->cfg()->block(loop_merge_block_id));
+    }
+  }
+}
+
+void StructuredLoopToSelectionReductionOpportunity::FixNonDominatedIdUses() {
+  // Consider each instruction in the function.
+  for (auto& block : *enclosing_function_) {
+    for (auto& def : block) {
+      if (def.opcode() == SpvOpVariable) {
+        // Variables are defined at the start of the function, and can be
+        // accessed by all blocks, even by unreachable blocks that have no
+        // dominators, so we do not need to worry about them.
+        continue;
+      }
+      context_->get_def_use_mgr()->ForEachUse(
+          &def, [this, &block, &def](Instruction* use, uint32_t index) {
+            // If a use is not appropriately dominated by its definition,
+            // replace the use with an OpUndef, unless the definition is an
+            // access chain in which case replace it with some (possibly fresh)
+            // global variable (as we cannot load from / store to OpUndef).
+            if (!DefinitionSufficientlyDominatesUse(def, use, index, block)) {
+              if (def.opcode() == SpvOpAccessChain) {
+                auto base_type = def.context()->get_type_mgr()->GetId(
+                    def.context()
+                        ->get_type_mgr()
+                        ->GetType(def.type_id())
+                        ->AsPointer()
+                        ->pointee_type());
+                use->SetOperand(index, {FindOrCreateGlobalVariable(base_type)});
```

Comment:
You may be changing the storage class. If the `use` is not an OpLoad or OpStore (may be another OpAccessChain) you may well have created invalid code by changing the storage class.

Reply:
Thanks!  I have generalized this functionality so that a dummy variable of a matching storage class is created, rather than always using private.  In the case of Function storage class it is added to the function, otherwise to global scope.

I am cautious about whether this will be OK for more interesting storage classes (e.g. related to input/output), and have opened an issue for this (#2183).